File: android/src/main/java/com/rnadmob/admob/RNAdMobAdHolder.java
package com.rnadmob.admob;

import android.util.SparseArray;

public class RNAdMobAdHolder<T> {
    SparseArray<T> adArray = new SparseArray<>();

    public void add(int requestId, T ad) {
        adArray.put(requestId, ad);
    }

    public T get(int requestId) {
        return adArray.get(requestId);
    }

    public void remove(int requestId) {
        adArray.delete(requestId);
    }

    public void clear() {
        adArray.clear();
    }
}

File: ios/RNAdMobAdHolder.swift
class RNAdMobAdHolder<T> {
    private var adArrayQueue = DispatchQueue(label: "RNAdMobAdHolderQueue")
    private var adArray = Dictionary<Int, T>()
    
    func add(requestId: Int, ad: T) {
        adArrayQueue.sync {
            adArray.updateValue(ad, forKey: requestId)
        }
    }
    
    func get(requestId: Int) -> T? {
        adArrayQueue.sync {
            return adArray[requestId]
        }
    }
    
    func remove(requestId: Int) {
        adArrayQueue.sync {
            adArray.removeValue(forKey: requestId)
        }
    }
    
    func clear() {
        adArrayQueue.async(flags: .barrier) {[weak self] in
            self?.adArray.removeAll()
        }
    }
}
